### PR TITLE
SAIC-491 Extend tests to verify one tenant is filtered.

### PR DESCRIPTION
**Pull request by plomakin**
Change filter files generation.
Also includes changes for this issues:
  SAIC-746 CF migrates admin's private images not included in filter.
  SAIC-485 Extend functional tests to verify migration succeeds when filtering by tenant is enabled and optimize_user_role_fetch is True
  SAIC-549 Extend functional tests to verify admin tenant with subnets migration
  SAIC-490 Extend functional tests in jenkins to verify migration with separate scenarios for resources and VMs

### CODE REVIEW

**Comment by plomakin** [comment]
Pls, merge gerrit gerrit commit first.

**Comment by MirantisWorkloadMobility** [comment]
Consider resolving pylint issues: \
``` \
************* Module cloudferrylib.os.actions.check_users
cloudferrylib/os/actions/check_users.py:46: [W0141(bad-builtin), CheckUsersAvailabilityOnSrcAndDst.run] Used builtin function 'map'. Using a list comprehension can be clearer.
cloudferrylib/os/actions/check_users.py:47: [W0141(bad-builtin), CheckUsersAvailabilityOnSrcAndDst.run] Used builtin function 'map'. Using a list comprehension can be clearer.
************* Module tests.test_vm_migration
devlab/tests/test_vm_migration.py:10: [E0203(access-member-before-definition), VmMigration.setUp] Access to member 'filter_vms' before its definition line 26
************* Module tests.test_resource_migration
devlab/tests/test_resource_migration.py:344: [W0212(protected-access), ResourceMigrationTests.test_migrate_tenant_quotas.get_tenant_quotas] Access to a protected member _info of a client class
devlab/tests/test_resource_migration.py:354: [W0212(protected-access), ResourceMigrationTests.test_migrate_tenant_quotas] Access to a protected member _info of a client class
************* Module tests.utils
devlab/tests/utils.py:130: [W0311(bad-indentation), ] Bad indentation. Found 20 spaces, expected 16
devlab/tests/utils.py:100: [W0621(redefined-outer-name), MigrationUtils.__init__] Redefining name 'config' from outer scope (line 17)
devlab/tests/utils.py:111: [E1129(not-context-manager), MigrationUtils.execute_command_on_vm] Context manager 'generator' doesn't implement __enter__ and __exit__.
************* Module tests.generate_load
devlab/tests/generate_load.py:20: [E0401(import-error), ] Unable to import 'exception'
devlab/tests/generate_load.py:878: [W0150(lost-exception), Prerequisites.create_tenant_wo_sec_group_on_dst] break statement in finally block may swallow exception

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

**Comment by MirantisWorkloadMobility** [comment]
Consider resolving pylint issues: \
``` \
************* Module devlab/tests/exception.py
devlab/tests/exception.py:1: [F0001(fatal), ] No module named devlab/tests/exception.py
************* Module cloudferrylib.os.actions.check_users
cloudferrylib/os/actions/check_users.py:46: [W0141(bad-builtin), CheckUsersAvailabilityOnSrcAndDst.run] Used builtin function 'map'. Using a list comprehension can be clearer.
cloudferrylib/os/actions/check_users.py:47: [W0141(bad-builtin), CheckUsersAvailabilityOnSrcAndDst.run] Used builtin function 'map'. Using a list comprehension can be clearer.
************* Module tests.test_vm_migration
devlab/tests/test_vm_migration.py:10: [E0203(access-member-before-definition), VmMigration.setUp] Access to member 'filter_vms' before its definition line 26
************* Module tests.test_resource_migration
devlab/tests/test_resource_migration.py:2: [W0403(relative-import), ] Relative import 'test_exceptions', should be 'tests.test_exceptions'
devlab/tests/test_resource_migration.py:344: [W0212(protected-access), ResourceMigrationTests.test_migrate_tenant_quotas.get_tenant_quotas] Access to a protected member _info of a client class
devlab/tests/test_resource_migration.py:354: [W0212(protected-access), ResourceMigrationTests.test_migrate_tenant_quotas] Access to a protected member _info of a client class
************* Module tests.utils
devlab/tests/utils.py:100: [W0621(redefined-outer-name), MigrationUtils.__init__] Redefining name 'config' from outer scope (line 17)
devlab/tests/utils.py:111: [E1129(not-context-manager), MigrationUtils.execute_command_on_vm] Context manager 'generator' doesn't implement __enter__ and __exit__.
************* Module tests.generate_load
devlab/tests/generate_load.py:20: [E0401(import-error), ] Unable to import 'test_exceptions'

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

File: cloudferrylib/os/actions/check_users.py [hunk 1]
```diff
@@ -28,6 +28,9 @@ class CheckUsersAvailabilityOnSrcAndDst(action.Action):
     """
 
     def run(self, **kwargs):
+        if self.cfg.migrate.migrate_users:
+            LOG.info("Users will be migrated. Skipping this check.")
+            return
```
Comment (roman-verchikov) [inline cloudferrylib/os/actions/check_users.py]: This seem unrelated to the patch, might be better to submit as a separate pull request

**Comment by roman-verchikov** [comment]
Please update pull request description and commit message to not have '...' in there.

File: devlab/tests/test_vm_migration.py [hunk 2]
```diff
@@ -26,19 +26,17 @@ def setUp(self):
         self.filter_vms = self.filtering_utils.filter_vms(src_vms)
 
     def test_vms_not_in_filter_stay_active_on_src(self):
-        filter_results = self.filter_vms
-        vms_filtered_out = filter_results[1]
         original_states = self.before_migr_states
-        for vm in vms_filtered_out:
+        for vm in config.vms_not_in_filter:
+            vm = [x for x in self.src_cloud.novaclient.servers.list(
+                search_opts={'all_tenants': 1}) if x.name == vm][0]
```
Comment (roman-verchikov) [inline devlab/tests/test_vm_migration.py]: Potential `IndexError`, please consider removing `[0]`

**Comment by roman-verchikov** [comment]
Looks good, other than few minor comments. @pavloivanets, @ntymtsiv please review

File: cloudferrylib/os/actions/check_users.py (hunk 1 above)
Comment (plomakin) [inline cloudferrylib/os/actions/check_users.py]: It's related to the patch, because we're using stage scenarios, that includes this check.

File: devlab/tests/test_vm_migration.py (hunk 2 above)
Comment (plomakin) [inline devlab/tests/test_vm_migration.py]: ok.

**Comment by MirantisWorkloadMobility** [comment]
Consider resolving pylint issues: \
``` \
************* Module devlab/tests/exception.py
devlab/tests/exception.py:1: [F0001(fatal), ] No module named devlab/tests/exception.py
************* Module cloudferrylib.os.actions.check_users
cloudferrylib/os/actions/check_users.py:46: [W0141(bad-builtin), CheckUsersAvailabilityOnSrcAndDst.run] Used builtin function 'map'. Using a list comprehension can be clearer.
cloudferrylib/os/actions/check_users.py:47: [W0141(bad-builtin), CheckUsersAvailabilityOnSrcAndDst.run] Used builtin function 'map'. Using a list comprehension can be clearer.
************* Module tests.test_vm_migration
devlab/tests/test_vm_migration.py:10: [E0203(access-member-before-definition), VmMigration.setUp] Access to member 'filter_vms' before its definition line 26
************* Module tests.test_resource_migration
devlab/tests/test_resource_migration.py:2: [W0403(relative-import), ] Relative import 'test_exceptions', should be 'tests.test_exceptions'
devlab/tests/test_resource_migration.py:344: [W0212(protected-access), ResourceMigrationTests.test_migrate_tenant_quotas.get_tenant_quotas] Access to a protected member _info of a client class
devlab/tests/test_resource_migration.py:354: [W0212(protected-access), ResourceMigrationTests.test_migrate_tenant_quotas] Access to a protected member _info of a client class
************* Module tests.utils
devlab/tests/utils.py:100: [W0621(redefined-outer-name), MigrationUtils.__init__] Redefining name 'config' from outer scope (line 17)
devlab/tests/utils.py:111: [E1129(not-context-manager), MigrationUtils.execute_command_on_vm] Context manager 'generator' doesn't implement __enter__ and __exit__.
************* Module tests.generate_load
devlab/tests/generate_load.py:20: [E0401(import-error), ] Unable to import 'test_exceptions'

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

**Comment by MirantisWorkloadMobility** [comment]
Consider resolving pylint issues: \
``` \
************* Module devlab/tests/exception.py
devlab/tests/exception.py:1: [F0001(fatal), ] No module named devlab/tests/exception.py
************* Module cloudferrylib.os.actions.check_users
cloudferrylib/os/actions/check_users.py:46: [W0141(bad-builtin), CheckUsersAvailabilityOnSrcAndDst.run] Used builtin function 'map'. Using a list comprehension can be clearer.
cloudferrylib/os/actions/check_users.py:47: [W0141(bad-builtin), CheckUsersAvailabilityOnSrcAndDst.run] Used builtin function 'map'. Using a list comprehension can be clearer.
************* Module tests.test_vm_migration
devlab/tests/test_vm_migration.py:10: [E0203(access-member-before-definition), VmMigration.setUp] Access to member 'filter_vms' before its definition line 26
************* Module tests.test_resource_migration
devlab/tests/test_resource_migration.py:2: [W0403(relative-import), ] Relative import 'test_exceptions', should be 'tests.test_exceptions'
devlab/tests/test_resource_migration.py:344: [W0212(protected-access), ResourceMigrationTests.test_migrate_tenant_quotas.get_tenant_quotas] Access to a protected member _info of a client class
devlab/tests/test_resource_migration.py:354: [W0212(protected-access), ResourceMigrationTests.test_migrate_tenant_quotas] Access to a protected member _info of a client class
************* Module tests.utils
devlab/tests/utils.py:100: [W0621(redefined-outer-name), MigrationUtils.__init__] Redefining name 'config' from outer scope (line 17)
devlab/tests/utils.py:111: [E1129(not-context-manager), MigrationUtils.execute_command_on_vm] Context manager 'generator' doesn't implement __enter__ and __exit__.
************* Module tests.generate_load
devlab/tests/generate_load.py:20: [E0401(import-error), ] Unable to import 'test_exceptions'

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

**Comment by pavloivanets** [comment]
Looks ok
+1

**Comment by plomakin** [comment]
Some of the functional test will expectedly fail or even error without changes from the gerrit repo.

File: devlab/tests/generate_load.py [hunk 3]
```diff
@@ -459,16 +463,20 @@ def update_filtering_file(self):
...
                 for vm in vm_id_list:
                     if vm != self.get_vm_id('not_in_filter'):
-                        vm_list.append(vm)
-                filter_dict[key]['id'] = vm_list
-        file_path = loaded_data[1]
-        with open(file_path, "w") as f:
-            yaml.dump(filter_dict, f, default_flow_style=False)
+                        filter_dict[key]['id'].append(str(vm))
+        for tenant in self.config.tenants + [{'name': 'admin'}]:
+            if 'deleted' in tenant and tenant['deleted']:
```
Comment (ntymtsiv) [inline devlab/tests/generate_load.py]: could be replaced with:
if tenant.get('deleted')

**Comment by MirantisWorkloadMobility** [comment]
Can one of the admins verify this patch?

**Comment by MirantisWorkloadMobility** [comment]
Can one of the admins verify this patch?

**Comment by MirantisWorkloadMobility** [comment]
Can one of the admins verify this patch?

**Comment by MirantisWorkloadMobility** [comment]
Can one of the admins verify this patch?

File: devlab/tests/generate_load.py (hunk 3 above)
Comment (plomakin) [inline devlab/tests/generate_load.py]: done

**Comment by MirantisWorkloadMobility** [comment]
Merge check failed
http://jenkins-cf.vm.mirantis.net:8080/job/cloudferry-unittests-and-style-checks/1726/

**Comment by MirantisWorkloadMobility** [comment]
Consider resolving pylint issues: \
``` \
************* Module devlab/tests/exception.py
devlab/tests/exception.py:1: [F0001(fatal), ] No module named devlab/tests/exception.py
************* Module cloudferrylib.os.actions.check_users
cloudferrylib/os/actions/check_users.py:46: [W0141(bad-builtin), CheckUsersAvailabilityOnSrcAndDst.run] Used builtin function 'map'. Using a list comprehension can be clearer.
cloudferrylib/os/actions/check_users.py:47: [W0141(bad-builtin), CheckUsersAvailabilityOnSrcAndDst.run] Used builtin function 'map'. Using a list comprehension can be clearer.
************* Module tests.test_vm_migration
devlab/tests/test_vm_migration.py:10: [E0203(access-member-before-definition), VmMigration.setUp] Access to member 'filter_vms' before its definition line 26
************* Module tests.test_resource_migration
devlab/tests/test_resource_migration.py:2: [W0403(relative-import), ] Relative import 'test_exceptions', should be 'tests.test_exceptions'
devlab/tests/test_resource_migration.py:344: [W0212(protected-access), ResourceMigrationTests.test_migrate_tenant_quotas.get_tenant_quotas] Access to a protected member _info of a client class
devlab/tests/test_resource_migration.py:354: [W0212(protected-access), ResourceMigrationTests.test_migrate_tenant_quotas] Access to a protected member _info of a client class
************* Module tests.utils
devlab/tests/utils.py:100: [W0621(redefined-outer-name), MigrationUtils.__init__] Redefining name 'config' from outer scope (line 17)
devlab/tests/utils.py:111: [E1129(not-context-manager), MigrationUtils.execute_command_on_vm] Context manager 'generator' doesn't implement __enter__ and __exit__.
************* Module tests.generate_load
devlab/tests/generate_load.py:20: [E0401(import-error), ] Unable to import 'test_exceptions'

**Comment by roman-verchikov** [comment]
+1, can be merged
